services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnRegister.java

```diff
@@ -0,0 +1,297 @@
+        Challenge challenge = new DefaultChallenge(context.getAuthenticationSession().getAuthNote(WebAuthnConstants.AUTH_CHALLENGE_NOTE));
+        ServerProperty serverProperty = new ServerProperty(origin, rpId, challenge, null);
+        // check User Verification by considering a malicious user might modify the result of calling WebAuthn API
+        boolean isUserVerificationRequired = policy.getUserVerificationRequirement().equals(WebAuthnConstants.OPTION_REQUIRED) == true ? true : false;
+
+        try {
+            WebAuthnRegistrationContext registrationContext = new WebAuthnRegistrationContext(clientDataJSON, attestationObject, serverProperty, isUserVerificationRequired);
+            // NOTE: not yet verify Attestation Statement based on certificates
+            WebAuthnRegistrationContextValidator webAuthnRegistrationContextValidator = WebAuthnRegistrationContextValidator.createNonStrictRegistrationContextValidator();
+            WebAuthnRegistrationContextValidationResponse response = webAuthnRegistrationContextValidator.validate(registrationContext);
+
+            showInfoAfterWebAuthnApiCreate(response);
+
+            checkAcceptedAuthenticator(response, policy);
+
+            WebAuthnCredentialModel credential = new WebAuthnCredentialModel();
+
+            credential.setAttestedCredentialData(response.getAttestationObject().getAuthenticatorData().getAttestedCredentialData());
+            credential.setAttestationStatement(response.getAttestationObject().getAttestationStatement());
+            credential.setCount(response.getAttestationObject().getAuthenticatorData().getSignCount());
+
+            this.session.userCredentialManager().updateCredential(context.getRealm(), context.getUser(), credential);
+
+            // store received Credential ID on Registration onto UserModel in order to be used on Authentication
+            String aaguid = response.getAttestationObject().getAuthenticatorData().getAttestedCredentialData().getAaguid().toString();
+            context.getUser().setSingleAttribute(WebAuthnConstants.PUBKEY_CRED_ID_ATTR, publicKeyCredentialId);
+            context.getUser().setSingleAttribute(WebAuthnConstants.PUBKEY_CRED_LABEL_ATTR, label);
+            context.getUser().setSingleAttribute(WebAuthnConstants.PUBKEY_CRED_AAGUID_ATTR, aaguid);
+            logger.infov("WebAuthn Registration successed. publicKeyCredentialId = {0}, publicKeyCredentialLabel = {1}, publicKeyCredentialAAGUID = {2}",publicKeyCredentialId, label, aaguid);
+
+            context.getEvent()
+                .detail("public_key_credential_id", publicKeyCredentialId)
+                .detail("public_key_credential_label", label)
+                .detail("public_key_credential_aaguid", aaguid);
+            context.success();
+        } catch (WebAuthnException wae) {
+            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
```

Comment:
What do you think about the following?

```java
        } catch (WebAuthnException wae) {
            if (logger.isDebugEnabled()) wae.printStackTrace();
            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
            return;
```

Reply:
I vote for:
```
        } catch (WebAuthnException wae) {
            if (logger.isDebugEnabled()) logger.debug(wae.getMessage(), wae.printStackTrace());
            setErrorResponse(context, ERR_WEBAUTHN_API_CREATE, wae.getMessage());
            return;
```

This will mean that exception will be logged to the logger, not to the error output like with "wae.printStackTrace()"

Reply:
Yes, I agree with you. I'll use logger.debug.